File: bytopic/binarytree/java/group1/InvertBinaryTree.java
// Given the root of a binary tree, invert the tree, and return its root.
// Example:
// Input: root = [4,2,7,1,3,6,9]
// Output: [4,7,2,9,6,3,1]
// Note: Visualize to understand
//
// Tag: 70/150
// Tag: 226/2927, R757/2936 (overall frequency ranking)

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class BinaryTreeInversion {
    public static void main(String...args) {
        TreeNode root = new TreeNode(4,
                new TreeNode(2, new TreeNode(1), new TreeNode(3)),
                new TreeNode(7, new TreeNode(6), new TreeNode(9))
        );

        TreeNode inverted = invertTree(root);
        System.out.print(inverted.val + "," + inverted.left.val + "," + inverted.right.val + ",");
        System.out.print(inverted.left.left.val + "," + inverted.left.right.val + ",");
        System.out.print(inverted.right.left.val + "," + inverted.right.right.val);
        System.out.println();
    }

    static TreeNode invertTree(TreeNode root) {
        if (root == null)
            return root;

        invertTree(root.left);
        invertTree(root.right);

        // Swapping
        TreeNode curr = root.left;
        root.left = root.right;
        root.right = curr;
        return root;
    }
}

File: bytopic/binarytree/python/group1/invert_binary_tree.py
"""
Given the root of a binary tree, invert the tree, and return its root.

Example:
Input: root = [4,2,7,1,3,6,9]
Output: [4,7,2,9,6,3,1]
Note: Visualize to understand

Tag: 70/150
Tag: 226/2927, R757/2936 (overall frequency ranking)
"""
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:  # Base case
        return root
    invert_tree(root.left)  # Call the left subtree
    invert_tree(root.right)  # Call the right subtree
    # Swap the nodes
    root.left, root.right = root.right, root.left
    return root


def main():
    root = TreeNode(4, TreeNode(2), TreeNode(7))
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(9)

    res = invert_tree(root)
    print(res.val, res.left.val, res.right.val, res.left.left.val, res.left.right.val,
          res.right.left.val, res.right.right.val)


if __name__ == "__main__":
    main()
